pysages/backends/core.py

```diff
@@ -42,18 +43,62 @@ def set_backend(name):
 
 
 def bind(context, sampling_method, **kwargs):
-    """Couples the sampling method to the simulation."""
-    #
+    """Couples the sampling method to the simulation.
+
+    context -- hoomd simulation context
+    sampling_method -- pysages sampling_method
+    callback -- callback method with call signature `callback(snapshot, state)`
+      called after PySages updated. Example: logging of CVs.
+    """
     if type(context).__module__.startswith("hoomd"):
         set_backend("hoomd")
     elif type(context).__module__.startswith("simtk.openmm"):
         set_backend("openmm")
-    #
+
     check_backend_initialization()
-    #
     return _CURRENT_BACKEND.bind(context, sampling_method, **kwargs)
 
 
 def check_backend_initialization():
     if _CURRENT_BACKEND is None:
         raise RuntimeError("No backend has been set")
+
+
+class ContextWrapper:
+    """
+    Abstract base class managing the context of the backends.
+    """
+    def __init__(self, context):
+        """
+        Store the context of the simulation to be wrapped.
+        """
+        self.context = context
+
+
+class Sampler:
+    """
+    Abstract base class to facilitate the update calls between backend and pysages.
+    """
+    def __init__(self, method_bundle, bias, callback=None):
+        """
+        Initialize the sampler.
+
+        method_bundle -- tuple with (snapshot, initialize, update) instances to be stored.
+        bias -- bias to be applied to the forces
+        callback -- optional callback function/functor with call signature (snapshot, state) called after updates. Examples: logging user calculated observables or CVs during the run.
+        """
+        snapshot, initialize, update = method_bundle
+        self.snapshot = snapshot
+        self.state = initialize()
+        self.update_from = update
+        self.bias = bias
+        self.callback = callback
+
+    def update(self):
+        """
+        Execute the update step.
+        """
+        self.state = self.update_from(self.snapshot, self.state)
+        self.bias(self.snapshot, self.state)
+        if self.callback:
+            self.callback(self.snapshot, self.state)
```

Comment:
```suggestion
            self.callback(self.snapshot, self.state, timestep)
```

Right now the `update` underlying C++ interfaces for HOOMD and OpenMM differ in that one takes a timestep (HOOMD) and the other doesn't, we should make the callback interface uniform regardless. See also a comment below.

Reply:
See comment above. I don't think we need an explicit timestep for the callback.

Reply:
So, I decided to pass the timestep now to the callback.
Maybe it helps a user, idk.

But that requires now, that OpenMM passes the timestep too for the update method